Driver Wrappers Codegen 1.0

silabs-ArchanaM opened the pull request:
## Description
This is the first and basic version of the CodeGen project tracked by #5137 .
It's in place just to clear us the structure of the folders, where code gen scripts will sit, basic python script to render the psa_cryto_driver_wrappers.c from replica template. 

## Status
READY

## Requires Backporting
 NO  

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

gilles-peskine-arm commented on .travis.yml:
```diff
@@ -20,6 +20,7 @@ jobs:
       python: 3.5 # version from Ubuntu 16.04
       install:
         - pip install mypy==0.780 pylint==2.4.4
+        - python3 -m pip install jinja2==2.10.1 types-jinja2
```
We're adding a new mechanism to manage the python module requirements in a single place, instead of spread around various CI scripts. See #5197. Please replace the commit "Jinja2 prerequisite set up" by rebasing on top of `development` once #5197 as been merged.

gilles-peskine-arm commented on scripts/make_generated_files.bat:
```diff
@@ -1,6 +1,11 @@
 @rem Generate automatically-generated configuration-independent source files
 @rem and build scripts.
 @rem Perl and Python 3 must be on the PATH.
+@rem psa_crypto_driver_wrappers.c needs to be generated prior to
```
Please update `tests/scripts/check-files.sh` as well.

gilles-peskine-arm commented on .travis.yml:
```diff
@@ -34,6 +35,7 @@ jobs:
       os: windows
```
Note: I'm working on adding support for extra python modules on FreeBSD and Windows on Jenkins. I'll let you know when that's done.

gilles-peskine-arm commented on scripts/codegen_driverwrappers/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,27 @@
+
+# write output to file
+outFile = open(driver_wrapper_output_filename,"w")
+outFile.write(result)
```
Complaint when building with `clang -std=c99 -pedantic -Wall -Wextra` (included on our CI with `tests/scripts/all.sh test_clang_opt`):
```
psa_crypto_driver_wrappers.c:2283:34: error: no newline at end of file [-Werror,-Wnewline-eof]
#endif /* MBEDTLS_PSA_CRYPTO_C */
```
Looks like jinja is stripping off the newline at the end of the last line?

gilles-peskine-arm replied:
https://github.com/ARMmbed/mbedtls/commit/f7a054bc07fb96ca9a8ebf5c6a64b6cde1cbd86c

silabs-ArchanaM replied:
Thanks for fixing this before I got around to !

gilles-peskine-arm commented on scripts/codegen_driverwrappers/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python3
+
```
Please move this script to the `scripts` directory, and its data files to `scripts/data_files`.

gilles-peskine-arm commented on scripts/codegen_driverwrappers/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,27 @@
+import os
+import jinja2
+
+def render(tpl_path):
+    path, filename = os.path.split(tpl_path)
+    return jinja2.Environment(
+        loader=jinja2.FileSystemLoader(path or './')
+    ).get_template(filename).render()
```
This would be more readable with intermediate variables. Something like:
```suggestion
def render(template_path: str) -> str:
        environment = jinja2.Environment(
        loader=jinja2.FileSystemLoader(os.path.dirname(template_path))
    )
    template = environment.get_template(os.path.basename(template_path))
    return template.render()
```

gilles-peskine-arm commented on library/Makefile:
```diff
@@ -302,6 +302,13 @@ version_features.c:
 	$(PERL) ../scripts/generate_features.pl
 
+psa_crypto_driver_wrappers.c: ../scripts/codegen_driverwrappers/generate_driver_wrappers.py
+psa_crypto_driver_wrappers.c:
```
Please add this file to `.gitignore`.

gilles-peskine-arm commented on library/Makefile:
```diff
@@ -302,6 +302,13 @@ version_features.c:
+psa_crypto_driver_wrappers.c: ../scripts/codegen_driverwrappers/generate_driver_wrappers.py
+psa_crypto_driver_wrappers.c:
+	echo "  Gen   $@"
+	$(PYTHON) ../scripts/codegen_driverwrappers/generate_driver_wrappers.py \
```
`PYTHON` isn't defined here, so this rule fails on platforms where the shebang line and the extension aren't enough to make the script executable. https://github.com/gilles-peskine-arm/mbedtls/commit/a17150d7acd44c5866a7117f25eca0b030e17a1e

gilles-peskine-arm commented:
On https://github.com/ARMmbed/mbedtls/commit/10e9a058c24de3ffba96f296b715e2c29b47afd8, all the Linux CI jobs are passing. The FreeBSD and Windows jobs will need https://github.com/ARMmbed/mbedtls/pull/5197.

gilles-peskine-arm commented:
https://github.com/ARMmbed/mbedtls/pull/5197 has been merged. Python dependencies are now tracked in a file within each branch and installed from CI scripts. Please rebase this pull request on top of the latest `development`.

No changes to `.travis.yml` or to `Dockerfile` should be needed anymore.

In https://github.com/ARMmbed/mbedtls/tree/dev/gilles-peskine-arm/codegen_1.0-pip I took part of this PR and adapted it, and I made sure that passed the CI. So this is probably a good basis for you to pick up.

silabs-ArchanaM commented:
> #5197 has been merged. Python dependencies are now tracked in a file within each branch and installed from CI scripts. Please rebase this pull request on top of the latest `development`.
> 
> No changes to `.travis.yml` or to `Dockerfile` should be needed anymore.
Dockerfile changes are still required, while running tests on the bionic docker.

gilles-peskine-arm commented:
The API check job should run normally now.

Github can't merge automatically (weird, a local `git merge` works), can you please rebase again on top of `development`?

silabs-ArchanaM commented:
I would prefer to move this into production as-is, what that helps with is we don't have to constantly rework with changes that are currently being made to the psa_crypto_driver_wrapper.c file. All subsequent changes will have to be made into the template file.
From rev 1.1 we can have a configurable guard till the templating is completely ready and merged.

silabs-ArchanaM commented on docs/proposed/psa-driver-wrappers-codegen-migration-guide.md:
```diff
@@ -28,5 +28,5 @@ While that is the larger goal, for version 1.0 here's what's changed
 
 (1) psa_crypto_driver_wrappers.c will from this point on be auto generated.
-(2) The auto generation is based on the template file at scripts/codegen_driverwrappers/psa_crypto_driver_wrappers.conf.
+(2) The auto generation is based on the template file at scripts/data_files/driver_templates/psa_crypto_driver_wrappers.conf.
```
A new directory for  driver wrapper templates was created under scripts/data_files. We expect the number of templates  to grow  so would make more sense to keep it organized in a folder and freeze the folder structure now.

minosgalanakis replied:
Would it make sense to rename that directory to something more indicative of the type of files it contains? Like code_templates, or something similar?

gilles-peskine-arm replied:
They are driver-related, so `driver_templates` is more precise than `code_templates`.

minosgalanakis commented on scripts/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,35 @@
+N = len(sys.argv)
+if N != 2:
+# This is the Root directory.
+    ROOT_DIR = ""
```
Considering that a script can be invoked as `python3 some_dir/scipt.py` or `cd some_dir && python3 script.py` it may be more portable to anchor it using the __file__ attribute since we already know where that script is relative to the project

~~~
# Python3 (not backwards compatbile)
import pathlib
pathlib.Path(__file__).parent.resolve()

# Python2 and 3
import os
os.path.dirname(os.path.abspath(__file__))
~~~

minosgalanakis commented on scripts/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,35 @@
+
+# Set template file name, output file name from the root directory
+DRIVER_WRAPPER_TEMPLATE_FILENAME = ROOT_DIR +\
+    "scripts/data_files/driver_templates/psa_crypto_driver_wrappers.conf"
```
It could be more portable to use the os.path.join facilities to compile the paths.

os.path.join("scripts", "data_files", "driver_templates", "psa_crypto_driver_wrappers.conf")

silabs-ArchanaM replied:
How is this more portable ?

minosgalanakis commented on scripts/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,35 @@
+
+def render(template_path: str) -> str:
+    environment = jinja2.Environment(
+        loader=jinja2.FileSystemLoader(os.path.dirname(template_path)),
```
This nested invocation can throw quite a few exceptions. It could be useful to catch them and notify the user as to why.

gilles-peskine-arm replied:
What would be the point of catching them? What notification could we offer that's better than Python's default trace?

gilles-peskine-arm replied:
Actually I do see one thing we should improve about error reporting, which is to indicate where Jinja is looking for files. This is the error when it can't find the template:
```
Traceback (most recent call last):
  File "/home/gilpes01/z/dev/mbedtls/reviewing/5193-codegen_1.0/library/../scripts/generate_driver_wrappers.py", line 30, in <module>
    RESULT = render(DRIVER_WRAPPER_TEMPLATE_FILENAME)
  File "/home/gilpes01/z/dev/mbedtls/reviewing/5193-codegen_1.0/library/../scripts/generate_driver_wrappers.py", line 13, in render
    template = environment.get_template(os.path.basename(template_path))
  File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 830, in get_template
    return self._load_template(name, self.make_globals(globals))
  File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 804, in _load_template
    template = self.loader.load(self, name, globals)
  File "/usr/lib/python3/dist-packages/jinja2/loaders.py", line 113, in load
    source, filename, uptodate = self.get_source(environment, name)
  File "/usr/lib/python3/dist-packages/jinja2/loaders.py", line 187, in get_source
    raise TemplateNotFound(template)
jinja2.exceptions.TemplateNotFound: psa_crypto_driver_wrappers.conf
```
Because we're passing the basename and the directory separately to Jinja, the error trace doesn't indicate the directory.

minosgalanakis commented on scripts/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,35 @@
+import os
+import jinja2
+
+def render(template_path: str) -> str:
```
I was wondering what is the purpose of setting the __annotations__ attribute at this point? Is that a requirement or a preference? 

As described in  the [PEP_483](https://www.python.org/dev/peps/pep-0483/) and [PEP_484 ](https://www.python.org/dev/peps/pep-0484/) specifications, my core concern is that it a very usefull but new feature, and would limit the possible systems where this script is deployed to anything running at least Python 3.5.

gilles-peskine-arm replied:
Type annotations are a preference of mine. We don't require them, but I strongly encourage them.

Python <3.6 is no longer supported upstream. We officially require Python ≥3.6 (see `README.md`). We still need support for Python 3.5 until we migrate our testing away from Ubuntu 16.04. Type annotations work from  Python 3.4 onwards. They're already used in several of our scripts.

gilles-peskine-arm commented on scripts/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,35 @@
+    template = environment.get_template(os.path.basename(template_path))
+    return template.render()
+
+N = len(sys.argv)
```
Not a blocker for now, but I will insist on it soon: this script should be importable, so command line parsing and side effects should only happen in a `main` function.

gilles-peskine-arm commented on scripts/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,35 @@
+
+N = len(sys.argv)
+if N != 2:
+# This is the Root directory.
```
“This” being the current directory. Clearer:
```suggestion
ROOT_DIR = os.curdir
```

gilles-peskine-arm commented on scripts/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,35 @@
+    ROOT_DIR = ""
+else:
+# Set the root based on the argument passed.
+    ROOT_DIR = sys.argv[1]
```
Passing the root directory as an argument doesn't make much sense to me. But ok for now. Later, the various file names should be passed as options (`import argparse`).

silabs-ArchanaM replied:
The idea of passing the root directory as an argument was to ensure we don't need to pass the arguments for file to be generated and template path for scripts from different places.
This is if  we know what  and where the template file is. So we just ensure we always run from the appropriate path with the ROOT_DIR being passed.

However, if we want to provide options for different template files to be used from custom paths, then yes, we will need to pass the arguments. 
Passing multiple files to the scripts in the way it is currently setup is honestly an eyesore .

gilles-peskine-arm replied:
The most common usage for the generator will be to pass a list of driver descriptions, and otherwise keep default settings. So I think these should be the positional arguments.
```
scripts/generate_driver_wrappers.py /path/to/my/accelerator.json /path/to/my/secure-element.json /path/to/third-party/secure-element.json
```

Passing alternative templates, output paths, etc., are secondary settings mostly useful for unusual build systems. I only see one use case for changing both the template path and the output path at once, and that's if you run the script from outside the source tree, e.g. to run the script from a different version of Mbed TLS, which is definitely not a common use case. So I don't see the point of `ROOT_DIR`.

silabs-ArchanaM replied:
I think we can rest this. Am rewriting this with argParse

silabs-ArchanaM replied:
> The most common usage for the generator will be to pass a list of driver descriptions, and otherwise keep default settings. So I think these should be the positional arguments.
> 

While driver descriptions are for the next version, am planning on listing the order of the drivers in a json, and not pass it as an argument. Based on the order a concatenated json would be created, parsed and templated in that order.

gilles-peskine-arm replied:
Having to create a file for the drivers doesn't sound convenient.

silabs-ArchanaM replied:
(Again, this is unrelated to this PR) 
But the way the build scripts are setup, one will need to pass it in multiple places.
Isn't it more convenient to have a driver list file which can be edited by the user  to specify the order of the driver jsons.
eg.,
]#cat driverlist.json
["mbedtls_test_opaque_driver.json","mbedtls_test_transparent_driver.json"] 

The user does not have to create this file, it will be pre created, but the user can edit it with their preferred order of drivers. Sounds good no ?

minosgalanakis replied:
Not a comment for this patch, but another trick you can could possibly do is add an indexing \__init\__.py file which parses the files, and creates a dictionary when invoked, and the caller script imports everything as a module.

\__init\__.py
~~~
def parse_directory():
     somelogic()

if __name__ == '__main__':
    driver_map = parse_directory()
~~~

somescript.py
~~~
from templates_dir import driver_map

for map_name, map_data in driver_map.items():
    ....

~~~

That way the scipt can use any level of iterators to list valid configs and even bind it with argparse outputs, without having to hardcode any logic or file list. In addition the benefit is that init file knows that it will always be in the directory of the templates.

gilles-peskine-arm commented on scripts/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,35 @@
+RESULT = render(DRIVER_WRAPPER_TEMPLATE_FILENAME)
+
+# Write output to file
+OUT_FILE = open(DRIVER_WRAPPER_OUTPUT_FILENAME, "w")
+OUT_FILE.write(RESULT)
+OUT_FILE.close()
```
Not a blocker for now:
```suggestion
with open(DRIVER_WRAPPER_OUTPUT_FILENAME, "w") as OUT_FILE:
    OUT_FILE.write(RESULT)
```
(And of course the variable name should be `out_file`, the only reason to call it `OUT_FILE` is that the code is not in a function so pylint insists on using the naming scheme for constants. Not a blocker for now per my other comment about a main function.)

gilles-peskine-arm commented on scripts/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,35 @@
+    ROOT_DIR = sys.argv[1]
+
+# Set template file name, output file name from the root directory
+DRIVER_WRAPPER_TEMPLATE_FILENAME = ROOT_DIR +\
+    "scripts/data_files/driver_templates/psa_crypto_driver_wrappers.conf"
```
Given an argument `foo`, this program will look for `fooscripts/data_files/driver_templates/psa_crypto_driver_wrappers.conf`, etc. 
```suggestion
DRIVER_WRAPPER_TEMPLATE_FILENAME = \
    os.path.join(ROOT_DIR, "scripts/data_files/driver_templates/psa_crypto_driver_wrappers.conf")
```

gilles-peskine-arm commented on scripts/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,35 @@
+
+# Set template file name, output file name from the root directory
+DRIVER_WRAPPER_TEMPLATE_FILENAME = ROOT_DIR +\
+    "scripts/data_files/driver_templates/psa_crypto_driver_wrappers.conf"
```
Why `.conf`? This isn't a configuration file. It should probably be called something like `psa_crypto_driver_wrappers.c.jinja`. The [Jinja documentation](https://jinja.palletsprojects.com/en/2.11.x/templates/#template-file-extension) suggests `.jinja`.

silabs-ArchanaM replied:
err... template configuration :). But am okay with .jinja.

gilles-peskine-arm commented on docs/proposed/psa-driver-wrappers-codegen-migration-guide.md:
```diff
@@ -0,0 +1,32 @@
+
+(1) psa_crypto_driver_wrappers.c will from this point on be auto generated.
+(2) The auto generation is based on the template file at scripts/data_files/driver_templates/psa_crypto_driver_wrappers.conf.
+(3) So while all driver wrapper templating support is yet to come in, the library user will need to patch into the template file as needed.
```
I would rather expect that if a user wants features that the generator doesn't support yet, they would write their own. This would be a transitional option, which we'd discontinue once the generator is sufficiently flexible.

silabs-ArchanaM replied:
If the user wants to change/ update something over the psa_crypto_driver_wrapper.c file they will need to patch the psa_crypto_driver_wrapper.conf template file. Changing all other makefiles to help generate this becomes unnecessary this way (I wouldn't want to make a segue for such presence). And there is no risk of the file being overwritten.

The intention of (3) here is that if a user wants to patch in some changes, that will need to be done over the template file. 
Once the feature is completely in lock, stock and barrel the template file will need very minimal inputs from us excepting cases where support for a new entry point is added. 

If this is not clear how would you like it reworded ? @gilles-peskine-arm

gilles-peskine-arm replied:
If I have a working C file, can I just replace the template with it? Or do some characters/words/… need escaping?

silabs-ArchanaM replied:
You could just replace the template file with the c file. 
The first rev 1.0 does the same, psa_crypto_driver_wrappers.conf is the same as psa_crypto_driver_wrappers.c.

gilles-peskine-arm reviewed (changes requested):
Blockers:

* Document how users can bypass the code generation if they need features that aren't supported yet, and ensure that the make/cmake code matches the documentation.
* A changelog entry needs to explain the changes for users who were editing `psa_crypto_driver_wrappers.c` directly. It can refer to the md document for details.
* The `ROOT_DIR` concatenation bug.

Not blockers for now: structure in the python code.

gilles-peskine-arm commented on scripts/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env python3
+"""This script is required for the auto generation of the
+   psa_crypto_driver_wrappers.c file"""
```
This will be called by build scripts. Typically from the makefile or cmakefile that we provide, but some users have their own build scripts.

silabs-ArchanaM replied:
For users who have their own buildscripts, this will need to be integrated just the way the other autogen files are like error.c  etc ?
Do we provide some specific guidance for such users ?

gilles-peskine-arm commented on library/CMakeLists.txt:
```diff
@@ -157,10 +157,26 @@ if(GEN_FILES)
             ${error_headers}
     )
+
+    add_custom_command(
```
I don't know what's wrong, but all out-of-source builds are failing. To reproduce:
```
rm -rf mbedtls_out_of_source_build; MAKEFLAGS=' ' ./tests/scripts/all.sh -k test_cmake_out_of_source
```

silabs-ArchanaM commented on library/psa_crypto_driver_wrappers.h:
```diff
@@ -1,7 +1,6 @@
 /*
  *  Function signatures for functionality that can be provided by
  *  cryptographic accelerators.
- *  Warning: This file will be auto-generated in the future.
```
I don't think this needs to be auto generated. That's the reason I removed it. 
Types defined by drivers need to be in their corresponding header files, included in the psa_crypto_driver_wrapper.c file via the json templating.

silabs-ArchanaM commented on library/Makefile:
```diff
@@ -318,6 +319,13 @@ version_features.c:
 
+psa_crypto_driver_wrappers.c: ../scripts/generate_driver_wrappers.py
+psa_crypto_driver_wrappers.c: ../scripts/data_files/driver_templates/psa_crypto_driver_wrappers.conf
+psa_crypto_driver_wrappers.c:
```
https://github.com/ARMmbed/mbedtls/pull/5193/files/947cf611f2f935701b2cef3333532b1c325e02bc#r772816254
Refer above comment.
If the user is doing either
(1) replacing the whole ....driver_wrapper.c or patching it, the user should do the same with the template file.  I see no problem in that approach.
(2)I understand it could be hard to visualize what the template file would look like after some templates are added in for each entry point. For that I recommend referring the prototype(prototype/driver_wrapper_sample.conf) at https://github.com/ARMmbed/mbedtls/pull/5067/commits/e705d07021f0aaebf8604651d24ff836975ca28e

daverodgman commented on scripts/generate_driver_wrappers.py:
```diff
@@ -1,28 +1,35 @@
-# render the template
-result = render(driver_wrapper_template_filename)
+# Set template file name, output file name from the root directory
+DRIVER_WRAPPER_TEMPLATE_FILENAME = ROOT_DIR +\
```
whitespace before `\`

daverodgman reviewed (approved):
Only some very minor style comments, otherwise looks good to me.

silabs-ArchanaM commented on scripts/codegen_driverwrappers/generate_driver_wrappers.py:
```diff
@@ -0,0 +1,27 @@
+        loader=jinja2.FileSystemLoader(path or './')
+    ).get_template(filename).render()
+
+n = len(sys.argv)
```
YEs rewriting with argparse, please expect a drop within the hour.

silabs-ArchanaM commented:
Restructured the Python script with argparse. Addressed all comments ( I think).
Updated ChangeLog and documentation reworded to confirm that the user can reuse the file they maintain as the template file in rev 1.0 as-is.

gilles-peskine-arm commented on scripts/generate_driver_wrappers.py:
```diff
@@ -1,35 +1,72 @@
+    Main with command line arguments.
+    """
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--mbedtls-root', nargs='?', default=None,
```
`mbedtls_root` only determines where to look for the templates. So why not have `--template-dir` instead?

silabs-ArchanaM replied:
parked to revisit in 1.1.

gilles-peskine-arm commented on ChangeLog.d/Driver_wrapper_codegen_1.0.txt:
```diff
@@ -0,0 +1,6 @@
+Features
+   * Add support for driver wrapper codegen 1.0. This autogenerates
+     the file library/psa_crypto_driver_wrappers.c. Rev 1.0 does not
+     support any additional templating. For further info on how to
+     patch into the psa_crypto_driver_wrappers.c refer
+     docs/proposed/psa-driver-wrappers-codegen-migration-guide.md
```
Please keep in mind that the target audience of this file is everyone who uses the library. Most readers will have never heard of “driver wrapper codegen 1.0”. The changelog entry should explain what changed. It should generally not mention things that haven't happened yet, although here it does make sense to explain the reason for the change (the file is generated now, because it will become variable later). Suggestion:

> The file library/psa_crypto_driver_wrappers.c is now generated from a template. In the future, the generation will support driver descriptions. For the time being, to customize this file, see docs/proposed/psa-driver-wrappers-codegen-migration-guide.md

Since there isn't yet any useful feature from the templating, this should go under “Changes” (to the build process) rather than “Features”.

We should revisit this changelog entry when 3.2 is released: depending on when the release happens, it may need to be consolidated with announcements of features that the generator does support by that point.

gilles-peskine-arm commented on scripts/generate_driver_wrappers.py:
```diff
@@ -1,35 +1,72 @@
+
+    with open(os.path.join(output_dir, "psa_crypto_driver_wrappers.c"), 'w') as out_file:
+        out_file.write(result)
+        out_file.close()
```
This line is redundant: `with open(…)` takes care of closing the file when it goes out of scope.

silabs-ArchanaM commented on library/Makefile:
```diff
@@ -320,11 +320,10 @@ version_features.c:
 	echo "  Gen   $@"
-	$(PYTHON) ../scripts/generate_driver_wrappers.py \
-                  "../"
+	$(PYTHON) ../scripts/generate_driver_wrappers.py --mbedtls-root .. .
```
Yes, just the output dir is enough !

silabs-ArchanaM commented on scripts/generate_driver_wrappers.py:
```diff
@@ -1,35 +1,72 @@
+    parser.add_argument('--mbedtls-root', nargs='?', default=None,
+                        help='root directory of mbedtls source code')
+    parser.add_argument('output_directory', nargs='?',
+                        default='library', help='output file\'s location')
```
That would mess it up when called from the root where you don't want to pass any argument, right  ? Am I missing something.

gilles-peskine-arm replied:
Er, no? When called from the root with no argument, root = current directory.

gilles-peskine-arm replied:
Or to put it in code:
```
$ cd library
$ ../scripts/generate_driver_wrappers.py
Traceback (most recent call last):
  File "../scripts/generate_driver_wrappers.py", line 74, in <module>
    sys.exit(main())
  File "../scripts/generate_driver_wrappers.py", line 69, in main
    generate_driver_wrapper_file(mbedtls_root, output_directory)
  File "../scripts/generate_driver_wrappers.py", line 50, in generate_driver_wrapper_file
    with open(os.path.join(output_dir, "psa_crypto_driver_wrappers.c"), 'w') as out_file:
FileNotFoundError: [Errno 2] No such file or directory: 'library/psa_crypto_driver_wrappers.c'
```
With `build_tree.guess_mbedtls_root`, the script can easily work out of the box in most common cases, without having to pass a different argument depending on the current directory.

silabs-ArchanaM replied:
I stand corrected ! Fixed.

silabs-ArchanaM commented on library/CMakeLists.txt:
```diff
@@ -164,11 +164,11 @@ if(GEN_FILES)
                 ${CMAKE_CURRENT_SOURCE_DIR}/../scripts/generate_driver_wrappers.py
-                ${CMAKE_CURRENT_SOURCE_DIR}/../scripts/data_files/driver_templates/psa_crypto_driver_wrappers.conf
-                ${CMAKE_CURRENT_BINARY_DIR}/psa_crypto_driver_wrappers.c
+                --mbedtls-root ${CMAKE_CURRENT_SOURCE_DIR}/..
```
But without this the out of source tree builds will fail.
